### ISSUE TITLE
Show decoded function & params on tx details v1

### ISSUE THREAD

**Issue by tschubotz** [gnosis/safe#329]
# What is this feature about? (1 sentence)
Show decoded function name + params on tx details.

# Why is it needed? What is the value? For whom do we build it?
It's the first step to add better tx decoding.

# High-level overview of the feature
**This feature should be aligned with the web interface.**
- The respective web epic is gnosis/safe-react#244

- Existing custom UI shouldn't be touch, i.e. for
  - Pure ETH transfers
  - Pure token transfers (ERC721/ERC20)
  - Settings changes

For other transactions, if the data is available from our backend:
- On the tx details, a user can see the called function and parameter names incl. parameter values, properly decoded as "Action"
- On the tx list, no changes are required right now.
- Multisends do not need to be decomposed. They just appear as generic contract interaction (i.e. calling the `multiSend` method on the `MultiSend` contract with 1 parameter that has a lot of data bytes.)

# Misc 

This is how this looks on web:
- Contract interaction:
![image (1).png](https://images.zenhubusercontent.com/57add4fee40e5714b16d4969/aadb408d-84d8-4e10-a689-c4fb44eb485f)

- Multisend:
![image.png](https://images.zenhubusercontent.com/57add4fee40e5714b16d4969/18f4f93d-3632-4376-a35c-42a03e5e6d5a)

# Tracking

- The screen view identifier should be `transaction_details_action` (cf. [screen view tracking docs](https://github.com/gnosis/safe/wiki/Overview-screen-tracking)

# Updates

- Tobi, August 20: Remove Multisend decomposing from this ticket and split out into separate one.
- Tobi, September 1: Add Screen tracking details

**Comment by DmitryBespalov** [comment]
Why is this called "Action" - is that OK for the users? From programmer's view point it is an **operation**, or a **function** that is called, or rather a **message** (per Solidity docs) that is sent to the receiving address.

**Comment by tschubotz** [comment]
"Action" is how we call it on the web interface so far, that's why I copied it here. Afaik there hasn't been any negative feedback about it, hence I think it's fair to keep.
@lukasschor pls correct me if I'm wrong.

**Comment by lukasschor** [comment]
Yes, also Etherscan uses "Action" in their wording

![image](https://user-images.githubusercontent.com/17196842/89183423-836da380-d597-11ea-92d2-ee9c5320765e.png)

**Comment by DmitryBespalov** [comment]
@tschubotz @lukasschor thanks!

I've roughly sketched the UI, does that look right to you?

![decoded data transactions 001](https://user-images.githubusercontent.com/284056/89191030-dba1a700-d591-11ea-858a-4f432eaa0f4d.jpeg)

**Comment by tschubotz** [comment]
@DmitryBespalov yes, I like it. Thanks for putting this together! 
I would use 
`Action: transferFrom` (i.e. colon instead of parentheses) to be a bit more in line with `Created:`, `Executed:` etc. further below.

**Comment by lukasschor** [comment]
![image](https://user-images.githubusercontent.com/17196842/89193965-0518fd80-d5a7-11ea-9c13-c4842aec6160.png)


If there's multiple actions we say "Acton 1 (...)", "Action 2 (...)" etc. on web

**Comment by lukasschor** [comment]
On web we call custom transactions "Contract Interactions". Not sure if this should be aligned as part of this ticket or a separate ticket (or not at all).

**Comment by tschubotz** [comment]
> On web we call custom transactions "Contract Interactions". Not sure if this should be aligned as part of this ticket or a separate ticket (or not at all).

"Contract interaction" is probably the nicer word here, agree.

**Comment by DmitryBespalov** [comment]
What if the multiSend contains 115 transactions (as we saw it is possible for the Gnosis Protocol transactions) - do we show 115 actions in the list or somehow make it on a separate screen or collapse them even more?

**Comment by MouazAlzahabi** [comment]
How about displaying the action name directly without the use of "Action" word?

**Comment by DmitryBespalov** [comment]
Also, in the previous app version, we were displaying  the multiSend's transactions the same as a transaction row from the transaction details - do we want to have that or the actions?

**Comment by tschubotz** [comment]
> How about displaying the action name directly without the use of "Action" word?

Could you please elaborate on why? Since Etherscan uses "Action" and our web interface uses "Action" I don't see a reason to do it differently, currently.



> What if the multiSend contains 115 transactions (as we saw it is possible for the Gnosis Protocol transactions) - do we show 115 actions in the list or somehow make it on a separate screen or collapse them even more?
>
> What would be the 'threshold' for displaying many transactions differently?

> Also, in the previous app version, we were displaying the multiSend's transactions the same as a transaction row from the transaction details - do we want to have that or the actions?

Good point. Collapsing a long list of items would not lead to a nice UX on mobile.
I tend towards the approach of the old app then to be honest, i.e. you would see "x transactions" or "Multisend with x transactions" and then after tapping on this, the user would see a list equivalent to the tx list where they can against open tx details on by one.

**Comment by DmitryBespalov** [comment]
@tschubotz I've started to make mock for the variant with a separate transaction list and then showing details of the transactions. I think this is not the way to go at this point, because this creates more work than needed. 

If we can live with the 115 actions in the transaction screen (edge case), that would be preferable. In that case we can reuse the existing UI and make the 'multiSend' screen all-in-one and reuse the collapsible actions that would be added for custom transactions screen.

Maybe we can move the "Actions" section to the bottom of the details - after the dates or in the Advanced section?

**Comment by DmitryBespalov** [comment]
Here's how the mocks look like
<img src="https://user-images.githubusercontent.com/284056/89290130-e9b1ff00-d647-11ea-9a2c-88a169e0272e.jpeg" height=1000px/>

**Comment by tschubotz** [comment]
> If we can live with the 115 actions in the transaction screen (edge case), that would be preferable. In that case we can reuse the existing UI and make the 'multiSend' screen all-in-one and reuse the collapsible actions that would be added for custom transactions screen.
> 
> Maybe we can move the "Actions" section to the bottom of the details - after the dates or in the Advanced section?

115 actions is really an edge case. Let's fix that later on. Then let's go with your original proposal. Of having the actions collapsed/expandible.
I would not move "actions" further down.

**Comment by MouazAlzahabi** [comment]
> Could you please elaborate on why? Since Etherscan uses "Action" and our web interface uses "Action" I don't see a reason to do it differently, currently.

Actually just for simplifying the UI, IMO it's clear enough that the method name is the action without the need to title it.

**Comment by lukasschor** [comment]
> > Could you please elaborate on why? Since Etherscan uses "Action" and our web interface uses "Action" I don't see a reason to do it differently, currently.
> 
> Actually just for simplifying the UI, IMO it's clear enough that the method name is the action without the need to title it.

 I would not introduce a discrepancy to the web app right now, without any user feedback to justify this. If at some point we get some insights on how to optimize this, we can switch mobile & web to a new approach in parallel. 

We don't know yet e.g. how important it is for users to quickly see the number of actions or to refer to a specific action using a number. So while it might seem "simpler" to just display method name, there might be some assumptions baked into that.

**Comment by posthnikova** [comment]
Designs are up for review, I didn't change much compared to Dima's version. 

**Non-multisend:** https://invis.io/6UWPBEPYZAP#/428290871_Transaction_Details_-Contract_Interaction-_

**Non-multisend expanded:** https://invis.io/6UWPBEPYZAP#/428290870_Transaction_Details_-Contract_Interaction-_Expanded-_

**Multisend:** https://invis.io/6UWPBEPYZAP#/428290869_Transaction_Details_-Contract_Interaction_-_Multisend-_

**Multisend expanded:** https://invis.io/6UWPBEPYZAP#/428290868_Transaction_Details_-Contract_Interaction_-_Multisend_Expanded-_

I agree that "batched" transaction list is not the right structure here. All that is unique for an action is its header, depending on the type of action. All other details (data, status, timestamp, confirmations, advanced, etherscan link) are shared. 'Internal transaction details' would double the info already present on the main transaction's screen.

**Comment by DmitryBespalov** [comment]
Looks good 👍

**Comment by tschubotz** [comment]
> Looks good 👍

I agree.

So far we have designs for parameter type address. 
What about adding examples for all other possible parameter types?

- Integers (uint): I think that's straight forward - we just display the number
- booleans: straight forward as well - display true or false
- data (bytes): that's a long string of hex data. how do we display that best?
- arrays: there can be arrays of any other type - how do we handle this? and array can be empty as well.

Did I miss anything>
For data and arrays I'm happy if we go for a simple solution rather than a perfect one. - how is this handled in the web interface? 🤔

**Comment by rmeissner** [comment]
While I like that we stick very closely to the web app design, I am not sure if this is the best approach here. There are some mobile UX considerations (collapsible list items vs opening a new screen) and there is also some potential small design iterations (e.g. making it more natural to read this information => On X Call Y (this is how I normally would like to read it). Obviously we need to pay attention on how much we change compared to the web version to not do something completely different :grimacing:

Also I would not take known shortcoming of the web UI into the mobile UI :smile: (e.g. if no data decoded is available in the multisend it shows no data at all ... not even the hex data) or lets say I would make it explicit that we want to implement the same shortcoming (basically declare it as a feature :P )

**Comment by jpalvarezl** [comment]
Carried over from slack: 

Looks good! Just to bring this to your attention (as I am not sure what is the scope of this first iteration) there is the possibility that some methods have array parameters for example this [transaction](https://safe-transaction.staging.gnosisdev.com/api/v1/safes/0x1230B3d59858296A31053C1b8562Ecf89A2f888b/creation/)
You will see that setup has _owners as an array type.
Again, not sure if this should be handled yet.

**Comment by posthnikova** [comment]
@tschubotz @jpalvarezl The simplest way to show all possible parameter types would be a list. There is also a version with icons and numbers to indicate this is an array: 

![image](https://user-images.githubusercontent.com/45787536/90141019-de16b480-dd7a-11ea-9681-118157408e8f.png)

As a follow up to refinement meeting I made two versions with nested navigation. The first has dropdowns, the second has a separate screen for each action (this is least preferred because you have to go back and forth):

![image](https://user-images.githubusercontent.com/45787536/90141091-f1298480-dd7a-11ea-8575-85b08c4f22bb.png)

**Comment by lukasschor** [comment]
For the nested txs, I would have imagined that the individual actions ("Action 1 (changeMastercopy)" etc.) are still listed in the first screen, but clicking on them would not expand the details, but rather open the details in a new screen.

**Comment by tschubotz** [comment]
> As a follow up to refinement meeting I made two versions with nested navigation. The first has dropdowns, the second has a separate screen for each action (this is least preferred because you have to go back and forth):

Thanks for preparing those versions! I would say that it's not a problem if they have to go back and forth to be honest.

**Comment by tschubotz** [comment]
> For the nested txs, I would have imagined that the individual actions ("Action 1 (changeMastercopy)" etc.) are still listed in the first screen, but clicking on them would not expand the details, but rather open the details in a new screen.

I thought the critique during backlog refinement was, that this first screen has too much info. Hence I think Sasha's option with the separate "multisend" list which resolves that.

**Comment by lukasschor** [comment]
> > For the nested txs, I would have imagined that the individual actions ("Action 1 (changeMastercopy)" etc.) are still listed in the first screen, but clicking on them would not expand the details, but rather open the details in a new screen.
> 
> 
> 
> I thought the critique during backlog refinement was, that this first screen has too much info. Hence I think Sasha's option with the separate "multisend" list which resolves that.

I understood the "too much info" part was mainly releated to the details of the transaction as they expand in the same view. And even having 3-4 actions expanded would result in a pretty long page.

I think it's still valuable to see the different method names already on the first page.

**Comment by rmeissner** [comment]
Personally having always exactly 1 entry on the details screen would be preferable. Therefore I really like Sashas proposal. As knowing that it is a multisig already provides me with some info that I can inspect further.

**Comment by sche** [comment]
I like the second option for the multisend. All the rest lgtm

**Comment by jpalvarezl** [comment]
Combining 1 level of navigation and then expanding seems reasonable. I am not sure how much nesting can be reasonably expected, but as a 1st iteration (or even as more long lasting solution), it looks pretty good 👍

**Comment by posthnikova** [comment]
Final designs:

**Non-multisend:** https://invis.io/6UWPBEPYZAP#/428290871_Transaction_Details_-Contract_Interaction-_

**Non-mutisend action details, all possible parameter types:** https://invis.io/6UWPBEPYZAP#/428290870_Transaction_Details_-Contract_Interaction-_Expanded-_

**Multisend:** https://invis.io/6UWPBEPYZAP#/428290869_Transaction_Details_-Contract_Interaction_-_Multisend-_

**Multisend actions list:** https://invis.io/6UWPBEPYZAP#/428290868_Transaction_details_-Multisend_Actions_List-_-2x

**Action details transfer:** https://invis.io/6UWPBEPYZAP#/429034148_Transaction_Details_-Multisend_Action_Details-_

**Change mastercopy:** https://invis.io/6UWPBEPYZAP#/429034174_Transaction_Details_-Multisend_Action_Details_2-_

**Set fallback handler:** https://invis.io/6UWPBEPYZAP#/429034210_Transaction_Details_-Multisend_Action_Details_3-_

**Comment by jpalvarezl** [comment]
Looks nice 👏 👏 !

I was wondering if it could be a good idea to make the title of the screen "Action (example)" (or simply what would be "example" in a real world example like "multiSend" or something else) and then we don't have to have the 1st line, saving up space by not needing to indent everything because of `Action (example)`
![Screenshot 2020-08-19 at 13 38 12](https://user-images.githubusercontent.com/11056031/90630593-9af88d80-e221-11ea-8313-6096f4cbfba4.png)

**Comment by jpalvarezl** [comment]
I think also, one of your past mocks, included the index of each value for array types. I don't know if it is important to keep that or not. I think it plays a role for arguments in method calls, but not sure if for parameters.

Here is the mock, I am talking about: 
![Screenshot 2020-08-19 at 13 43 28](https://user-images.githubusercontent.com/11056031/90630891-21ad6a80-e222-11ea-8e6b-8e3eb1cd2e15.png)

**Comment by posthnikova** [comment]
@jpalvarezl On the one hand I wanted 'Action 1, Action 2 etc' to remain so it corresponds to the web interface. On the other, without indentation it looks nicer. `Action (example)` in the nav bar would be too long. 

![image](https://user-images.githubusercontent.com/45787536/90640257-106b5a80-e230-11ea-9a66-d66373a94ad8.png)

> I think also, one of your past mocks, included the index of each value for array types

Additional formatting would be harder to implement so I didn't go forward with this version.

**Comment by jpalvarezl** [comment]
That looks great with no indentation and the method name in the title bar. Worst case scenario, we can ellipsize it :)

**Comment by liliya-soroka** [comment]
@tschubotz and @posthnikova are these links on the design for the v1? :
https://app.zeplin.io/project/5e8f27b9963e55bb79b13634/screen/5eb2935f59a84577c2909c17 
https://app.zeplin.io/project/5e8f27b9963e55bb79b13634/screen/5f4e3aa7f98a800e851a8bfd

**Comment by posthnikova** [comment]
> are these links on the design for the v1?

@liliya-soroka Yes, this should be in v1

**Comment by liliya-soroka** [comment]
Verified
app version:
ios : 2.3.0 (161)
android: 2.3.0 - 1278rc
